### PR TITLE
Incremental compact index

### PR DESCRIPTION
**Pull request by skaes**
This PR build on the work of @ahorek (https://github.com/geminabox/geminabox/pull/353). To the best of my knowledge, it completely implements the bundler compact index.

It fixes some problems of the original PR:

1. The compact index is now built incrementally, reducing the upload time of new gem versions from several minutes to a few seconds.
2. It correctly implements gem deletion.
3. It fixes race conditions in the original PR which could cause incomplete information to be server by the API
4. It implements the /names API for proxy mode

On top of that it moves most of the indexing code out of server.rb which makes that file way more readable and allows for more unit testing of the indexing related code.

While functionally complete with acceptable performance, there are still a number of open questions

1. The plethora of configuration options makes it really hard to get good test coverage. It is also confusing for users.
2. Shipping unsafe configuration options feels wrong to me.
3. Running a proxy server is basically broken due to adding the remote gem to the local index (see https://github.com/geminabox/geminabox/pull/345).
4. Gem deletion is still inefficient (takes minutes on our server, I plan to address this in a separate PR)
5. Test coverage of the compact index should probably be improved.

My suggestions are:
1. Drop the incremental_updates option, it should always be on.
2. Drop the combine_local_and_remote_gem_versions strategy. The compact index API is essential for bundler to run efficiently, but I see no way to implement it with this strategy enabled.
3. Drop the compact_index option, it should always be on as bundler more or less expects it. Yes, it does have a fallback, but bundle install operations without Gemfile.lock or bundle updates when it's already there become rather slow.
4. Drop the remote_gems_take_precedence_over_local_gems strategy as it is completely unsafe.

In my mind there are currently only two viable configurations: standalone server and caching proxy without local gems. As soon as you mix local gems, the issue described https://github.com/geminabox/geminabox/pull/345 strikes and renders proxying ineffective.

@tnir @ahorek @deivid-rodriguez please have a look. I would very much appreciate your opinion on the open questions.

### CODE REVIEW

**Comment by skaes** [comment]
While we're on the topic of dropping old things: why do we test against old, insecure end of life'd Ruby versions and the totally obsolete 1.8.25 rubygems versions? Can we drop those tests?

**Comment by ahorek** [comment]
@skaes thank you, it looks great! I agree with your suggestions.

Note that these proxy strategies were also a common source of security issues and the new API simply prevents using them. I don't personally need them and I think there are other options to achieve the same goal safer...

@tnir ? what do you think?

**Comment by skaes** [comment]
I went ahead and implemented incremental index updates for gem deletion. This means that a full reindex will only be necessary when the index has been corrupted for whatever reason.

I have improved test coverage as well, which was much simpler than before due to having the indexer logic extracted from the server class.

I have also fixed some minor problems that are still on master, for example incorrect test  coverage reports.

@tnir could you please weigh in on the open questions and let me know how to proceed?

Thx a lot.

**Comment by tnir** [comment]
@skaes Thanks for an amazing improvement!

> This PR build on the work of @ahorek (https://github.com/geminabox/geminabox/pull/353). To the best of my knowledge, it completely implements the bundler compact index.

I thought #353 can address the original problem, so I am not sure if we should prefer #435 over #353, which I might need to have more time to understand.

> I have also fixed some minor problems that are still on master, for example incorrect test coverage reports.

Thank you too for such improments. If any of them is independent from index API implementation, could you create separate PR(s)? 🙏

**Comment by ahorek** [comment]
@tnir #435 should be preferred, The API is still the same, but @skaes improved my original proposal in many ways. Thank you again for your work!

**Comment by skaes** [comment]
@tnir I have extracted the test setup changes to a separate PS https://github.com/geminabox/geminabox/pull/476.

**Comment by skaes** [comment]
@tnir can you please [this code climate check](https://codeclimate.com/github/geminabox/geminabox/pull/435)? I seems to expect a Rails app, which we do not have.

**Comment by tnir** [comment]
@skaes Sorry for that. When I updated configuration for the plugins, I made a mistake 🙇 Removed brakeman plugin 🙇

**Comment by tnir** [comment]
@skaes Sorry for that again. As I just found brakeman was still enabled 🤔 , disabled brakeman plugin again 🙇

**Comment by skaes** [comment]
@tnir the code climate run failed and it looks like that's a config problem:

```
We had trouble running the grep engine.
The engine's output shown below may indicate the cause of the error.
If not, please [contact us](https://codeclimate.com/help) so we can investigate further.

key not found: "config"
```

**Comment by skaes** [comment]
@tnir @ahorek @deivid-rodriguez I believe the PR is now complete. It implements the compact index api for standalone servers and proxy servers. I have added[ a script](https://github.com/skaes/geminabox/blob/c759dc9c4887dd81cd2a842bcb92ee6c91441a5b/README.md#upgrading-from-older-geminabox-versions) to convert existing proxies to the new disk layout. Reviews would be great. I'm going to update our own in-house server to the new code this week.

File: lib/geminabox/dependency_info.rb
Comment (ahorek) [inline lib/geminabox/dependency_info.rb:90]: since ruby 2.2 support has been dropped, we can use
```suggestion
      str = +PREAMBLE
```

File: lib/geminabox/remote_cache.rb
Comment (ahorek) [inline lib/geminabox/remote_cache.rb:41]: cache entries are saved in binary mode, it should be safer to read them the same way
```suggestion
      File.binread(cache_path + file_path)
```

File: lib/geminabox/dependency_info.rb
Comment (skaes) [reply lib/geminabox/dependency_info.rb:90]: Rubocop does not like it, that's why I changed it.

File: lib/geminabox/remote_cache.rb
Comment (skaes) [reply lib/geminabox/remote_cache.rb:41]: thx!

File: lib/geminabox/dependency_info.rb
Comment (skaes) [reply lib/geminabox/dependency_info.rb:90]: Also, I really did not know this pattern before I saw it in your code.

File: lib/geminabox/dependency_info.rb
Comment (skaes) [reply lib/geminabox/dependency_info.rb:105]: As above, forbidden by rubocop and somewhat cryptic.

File: lib/geminabox/dependency_info.rb
Comment (ahorek) [reply lib/geminabox/dependency_info.rb:90]: ok, I see, it won't make a difference here anyway...

File: lib/geminabox/compact_indexer.rb
Comment (ahorek) [inline lib/geminabox/compact_indexer.rb:136]: FYI threads in CRuby still do use GIL, so it won't speed up CPU-bound calculations, it could speed up IO-bound operations like writing to a filesystem.
using non-blocking operations like IO.read_nonblock or async should be a better way how to improve IO performance, but I don't think there's a gem with a simple interface like the Parallel gem. At least the maximum number of threads is limited, so I agree this is probably the simplest way to improve it.

Comment (skaes) [reply lib/geminabox/compact_indexer.rb:136]: Yes the GIL is a problem I had almost forgotten about. In fact, the speedup here isn't all that impressive (6 instead of 7 minutes on my laptop for our gem server). But I converted a lot of time back and forth between proxy and standalone to test my conversion script and it helped a bit.

**Review by ahorek — APPROVED** [review]
👍

File: lib/geminabox.rb
Comment (skaes) [reply lib/geminabox.rb:14]: Pathname is part of the stdlib, I wouldn't really count this as an extra dependency, at least not a complicated one. That said: the code is littered with File.join operations and the knowledge of directory structure is dispersed all over the code base. That's much worse. I meant to centralize this and use pathnames everywhere. But left it for another PR.

**Comment by skaes** [comment]
Converted PR to draft again as I discovered a few problems while going live on our server. I'm investigating.

**Comment by tnir** [comment]
Thank you.

> but the vast majority of the 251 complaints is just nonsensical. For starters, please exclude CHANGELOG.md and README.md.

@skaes Updated the problems existing in the existing code in another recent PR.

**Comment by jeremy** [comment]
Great work @skaes. 🙇🏼‍♂️

**Comment by skaes** [comment]
@tnir I think the PR is ready to merge now. We have been running the code in production in standalone mode for over a week now and I have tested the proxy functionality with a copy of our production gem server data.

I tried to fix all the Code Climate complaints, but there are three left which I don't intend to fix, because all fixes I could think of would just obscure the code IMHO.

**Comment by tnir** [comment]
@geminabox/maintainers Could you mind reviewing this change? cc @skaes

**Comment by tnir** [comment]
@skaes Thanks. I passed it to maintainers as I am not inclined to do this so much...

**Comment by keithrbennett** [comment]
This issue has been a problem with my code base as well, so I look forward to the release of this fix. Thanks to all of you for your efforts.

**Comment by triskweline** [comment]
We're running a larger gem server using Geminabox. Having the compact index endpoint would speed up Bundler runs considerably. Hoping this can be merged!

**Comment by ahorek** [comment]
@tnir any news? unfortunately, there seems to be no feedback from other maintainers...

**Comment by skaes** [comment]
@ahorek @triskweline I think this is going nowhere.

The only significant changes on master have been the introduction of rubocop, code climate and a code of conduct.

At this point  I am considering a fork or moving our server to something else entirely.

**Comment by allcentury** [comment]
@skaes can you fix the merge conflict and ensure CI is passing?  

I haven't contributed to this in a long time but as long as @geminabox/maintainers have no objections, I can add you as a maintainer so we can merge this and you can help with any bug fixes related to these changes.  

thanks for your contributions!

**Comment by skaes** [comment]
Hi @allcentury, thx for the offer. There's more to be done than just merging this PR. My experience with running the compact index on production lead to a few more changes, mostly to ensure acceptable performance (see https://github.com/skaes/geminabox/pull/42).

Completing the PR will be quite some effort (rebasing, documentation) and I don't want to spend it in vain only for the PR to linger here.

I'd be willing to help maintain the project, but I'm not sure how we can work together effectively, if no one but me is willing or able to review larger changes that I'm sure would lie ahead of us if I would join the group.

**Comment by allcentury** [comment]
@skaes are you looking for beta users for this before merging?  This PR is very large so I don't see how anyone will be able to thoroughly review and test this without using it in their own environments.  The other option is we could identify this as a module in alpha/beta and guard against it until we have more users/confidence.  

This project is open source so there is no authority here other than a collective union of volunteers.  Let me know how I can help you proceed

**Comment by conathan** [comment]
(responding to the no-pr-activity-tag), but this is still not resolved, so posting as requested by the github-actions bot.

Without it, bundler takes a long time calculating dependencies, and I believe there was even a version of bundler that removed using the dependency api (but then later readded, probably due to projects like geminabox that do not have patches like this done yet.

Speaking of, been using an earlier version of this patch for a long time as newer versions of ruby were almost unuseable with geminabox (Slow, I believe high ram as well but been a long time).  Production environment with multiple ruby on rails environments (as well as using cinc-client, capistrano, via rvm).
* Originally was using the patches from #353, before moving onto this

We use bundler version 1.7.8, with ruby 1.8.7,  as well as bundler 1.16.1 with ruby 2.0/2.5  (as well as ruby 3.1.3 installed via rvm, managed via gem install).  No known issues, other then bundler 1.7.8/ruby2.0 not figuring out what rake to install in a Gemfile (something I forever, prior to this patch), which went away when I bumped the ruby 2.0 bundler version to 1.16.1 (so I consider it a non issue)

Would love to see this committed, and move onto it, I could see if I can rebase our patch set on the latest changes if that would be of any help  (Current patch was done by using something like https://patch-diff.githubusercontent.com/raw/geminabox/geminabox/pull/435.patch, and then trying to apply it to the gems installed via bundle install
* Gemfile has gem parallel, and ruby-progressbar in it, as we install the 2.1.0 version prior to patching it.

**Comment by gatopanx** [comment]
FYI 5 days left https://blog.rubygems.org/2023/04/07/dependency-api-deprecation-delayed.html

**Pull request closed without merging**
